crates/compose/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+            .with_context(|| "failed to encode composition graph")
+            .map(Into::into)
+    }
+
+    fn new(loader: &'a L) -> Self {
+        Self {
+            graph: CompositionGraph::new(),
+            loader,
+        }
+    }
+
+    async fn prepare_dependencies(
+        &mut self,
+        world_id: WorldId,
+        component: &LockedComponent,
+    ) -> anyhow::Result<IndexMap<String, DependencyInfo>> {
+        let mut imports = self.graph.types()[world_id].imports.clone();
+
+        let mut prepared = IndexMap::new();
+
+        // TODO: collect and return conflicts
+
+        for (dependency_name, dependency) in component.dependencies.iter() {
+            // As we match dependency names against import names, we remove the
+            // matched import name from the imports map to prevent duplicate
+            // matches. Collecting the import keys here is an optimization to
+            // avoid walking the full set of import keys on each iteration.
+            let import_keys = imports.keys().cloned().collect::<Vec<_>>();
+
+            let mut matched = Vec::new();
+
+            for import_name in import_keys.iter() {
+                if matches_import(dependency_name, import_name) {
+                    match imports.swap_remove(import_name) {
+                        Some(_item) => {
+                            matched.push(import_name.clone());
+                            break;
+                        }
+                        None => {
+                            anyhow::bail!(
+                                "import '{import_name}' of component '{component_id}' is already used by another dependency",
+                                component_id = component.id,
+                            );
+                        }
+                    }
+                }
+            }
+
+            if matched.is_empty() {
+                anyhow::bail!(
+                    "dependency '{dependency_name}' doesn't match any imports of component '{component_id}'",
+                    component_id = component.id,
+                );
+            }
+
+            let info = self
+                .register_dependency(dependency_name.clone(), dependency)
+                .await?;
+
+            // Insert the expanded dependency name into the map detecting duplicates
+            for import_name in matched.into_iter() {
+                if let Some(previous_info) = prepared.insert(import_name.clone(), info.clone()) {
+                    anyhow::bail!(
+                        "component '{component_id}' dependency '{dependency_name}' introduces import '{import_name}' which conflicts with dependency '{previous_dependency_name}'",
+                        component_id = component.id,
+                        previous_dependency_name = previous_info.manifest_name,
+                    );
+                }
+            }
+        }
+
+        Ok(prepared)
+    }
+
+    async fn build_instantiation_arguments(
```

Comment:
These functions could very much use a bit of explanation of what they are used for.

Reply:
Updated, PTAL.